File: Ch7/7_2_Fence.cpp
// Q : from fence made out of N woods, find the size of the largest reactange

// Idea : Brute Force
// find all possible reactangles and figure out the max size

#include <iostream>
#include <vector>

using namespace std;

int main(){
    // input test cases
    int c;
    cin >> c;
    for(int i = 0; i < c; i++){
        // input number of woods
        int n;
        cin >> n;

        // input heights of woods
        int input;
        vector<int> heights;
        for(int j = 0; j < n; j++){
            cin >> input;
            heights.push_back(input);
        }

        int max = 0;
        for(int j= 0; j < n; j++){
            int length = 1;
            // find left side length
            for(int left = j - 1; left >= 0 && heights[left] >= heights[j]; left--){
                length++;
            }

            // find right side length
            for(int right = j + 1; right < n && heights[right] >= heights[j]; right++){
                length++;
            }

            // calculate area
            int area = length * heights[j];

            // save the largest area
            if(area > max)
                max = area;
        }
        cout << max << endl;
    }
}

File: Ch7/7_2_Fence.py
# Q : from fence made out of N woods, find the size of the largest reactange

# Idea : Divide and Conquer
# divide the fence in half
# the largest should be either in the left, in the right, or sideded in the middle


def solve(left, right):
    # base case : left == right
    if left == right :
        return heights[left]
    
    half = (left + right) // 2

    # check largest in the left and right
    res = max(solve(left, half), solve(half + 1, right))

    # check largest in the middle
    low = half
    high = half + 1
    
    # check smallest middle position rectangle (size 2)
    height = min(heights[low], heights[high])
    res = max(res, height * 2)

    # iterate until cannot expand to left and right
    while left < low or high < right :
        # expand to more larger height
        # expand to right
        if high < right and (low == left or heights[low - 1] < heights[high + 1]):
            high += 1
            height = min(height, heights[high])
        # expand to left
        else :
            low -= 1
            height = min(height, heights[low])
        
        # compare area
        res = max(res, height * (high - low + 1))

    return res


# input test cases
c = int(input())
for i in range(c):
    # input number of woods
    n = int(input())
    heights = list(map(int, input().split()))
    print(solve(0, n - 1))

# O(n) = nlogn algorithm